packages/ipfs/src/cli/commands/dag/import.js

```diff
@@ -0,0 +1,73 @@
+'use strict'
+
+const fs = require('fs')
+const CID = require('cids')
+const CAR = require('datastore-car')
+const Block = require('ipfs-block')
+const { cidToString } = require('../../../utils/cid')
+
+module.exports = {
+  command: 'import [carfile...]',
+
+  describe: 'Import the contents of CAR files from disk or stdin.',
+
+  async handler ({ ctx, carfile }) {
+    const { ipfs, print, getStdin } = ctx
+    let count = 0
+
+    if (carfile) { // files
+      for await (const file of carfile) {
+        print(`importing from ${file}...`)
+        count += await importCar(ipfs, print, fs.createReadStream(file))
+      }
+    } else { // stdin
+      print('importing CAR from stdin...')
+      count = await importCar(ipfs, print, getStdin())
+    }
+
+    print(`imported ${count} blocks`)
+  }
+}
+
+async function importCar (ipfs, print, inStream) {
+  const car = await CAR.readStreaming(inStream)
+  const roots = await car.getRoots()
+  const rootStatus = roots.reduce((p, cid) => {
+    p[cid.toString()] = false
+    return p
+  }, {})
+  let count = 0
+
+  // CAR.readStreaming acts like a datastore that we can streaming query()
+  for await (const { key, value } of car.query()) {
+    // key is the cid as a string and value is the binary block data
+    const cid = new CID(key)
+    const block = new Block(value, cid)
+    await ipfs.block.put(block)
+    if (rootStatus[key] !== undefined) {
+      // TODO: what if it's already true? double block in the CAR!
+      rootStatus[key] = true
+    }
+    count++
+  }
+
+  // A well-formed CAR will have roots that reference blocks that actually exist
+  // in the CAR body, but that may not be the case. Don't try to pin if the root
+  // didn't exist in the body.
+  // TODO: ^ go-car currently attempts to pin roots even if they don't exist in
+  // the CAR body, need to align behaviour
```

Comment:
@rvagg this is done in `go-ipfs` to allow a copy-less "transactional" operation:
```
<some source> | stream-dagger <many options> --emit-stdout=car-v0-fifos-xargs | xargs -0 ipfs dag import
```
What that mode does is print 2 fifo names on stdout. The first fifo contains all the data. The second contains the roots only ( because we can derive the roots only once we streamed everything over ). The full `dag import` context serves as a "transaction" of sorts, keeping GC at bay between the lengthy data stream and the pin at the very end.

Reply:
i’d wait to support this until a later date if `js-ipfs` wants to prioritize it.